projects/storefrontlib/cms-components/product/product-list/container/product-list-component.service.ts

```diff
@@ -84,10 +85,26 @@ export class ProductListComponentService {
           state.params,
           state.queryParams
         );
-        this.search(criteria);
+        this.searchIfQueryHasChanged(criteria);
       })
     );
 
+  /**
+   * Search only if the previous search query does NOT match the new one.
+   * This prevents repeating product search calls for queries that already have loaded data.
+   */
+  protected searchIfQueryHasChanged(criteria: SearchCriteria) {
+    this.productSearchService
+      .getResults()
+      .pipe(take(1))
+      .subscribe((results) => {
+        const loadedQuery = results?.currentQuery?.query?.value;
+        if (loadedQuery !== criteria.query) {
+          this.search(criteria);
```

Comment:
[MAJOR]
I did QA now and I can see the pagination is not working anymore.
I believe that we should check not only for `results.currentQuery` but also for `results.pagination`and compare it with our `SearchCriteria` (pageSize,currentPage,sort)

Reply:
I realized only now, that even our fixed implementation covering also `pageSize,currentPage,sort` might be a breaking change - it won't take into account any other custom URL params that customers might have created and used in their customized search mechanism.

That said, I think we should cover our bugfix with a feature flag.

Reply:
These issues should be fixed now. I had to add quite few different methods to handle each case unfortunately because the handling for each situation is non-uniform (as you will be able to tell by the functions I added).